### ISSUE TITLE
API to get the Unique Device Id 

### ISSUE THREAD

**Issue by aghassemi** [flutter/flutter#326]
It would be nice to provide an API to get the Unique Device Id in Flutter.

@abarth

**Comment by abarth** [comment]
Can you add a link to the exact Android API that you'd like to be made available?

**Comment by abarth** [comment]
How persistent does the unique identifier need to be?  For example, does it need to survive restarts of the app?  Restarts of the phone?  Reinstall of the app?  Flashing a new OS?  Factory reseting the device?

**Comment by aghassemi** [comment]
I was thinking to get the ANDROID_ID (Settings.Secure.ANDROID_ID) or something equivalent to that. Android Id seems to be generated first time device boots. It is reset when the device is wiped.

@abarth I removed the slides label as this is no longer needed by the slides app since we are now using a different mechanism to uniquely identify the service (Syncbase) running on the device. ( We now rely on discovery and an service address - normally IP:Port:Vanadium Generated Unique ID )

This might still be a useful feature to have in Flutter but not needed by slides anymore.

**Comment by abarth** [comment]
Thanks!

**Comment by stevenroose** [comment]
This is crucial. A lot of apps rely on some kind of unique device identifier.

Android provides one here: https://developer.android.com/reference/android/provider/Settings.Secure.html#ANDROID_ID

So it would be useful is Flutter provided a way to access that same identifier.

**Comment by abarth** [comment]
Got it.

**Comment by Hixie** [comment]
What would this be used for? Surely if you want to identify the user, it's better to ask the user for an identifier (e.g. have a login system), no?

**Comment by abarth** [comment]
@Hixie Philosophically, I think we should have an "android_services" package that uncritically reflects all the services that exist in Android.  For this issue, we should probably reflect all of these APIs:

https://developer.android.com/reference/android/provider/Settings.Global.html
https://developer.android.com/reference/android/provider/Settings.Secure.html
https://developer.android.com/reference/android/provider/Settings.System.html

Obviously, the "android_services" won't be available on iOS or FNL.  You'll need "ios_services" for iOS, and for FNL you'll be able to use the system-provided Mojo interfaces.

To keep ourselves straight, we should aim to make "android_services" an optional module that isn't tightly coupled with the rest of Flutter, much in the same a third-party developer would make a module that wraps a binary SDK.

**Comment by Hixie** [comment]
seems reasonable.

**Comment by stevenroose** [comment]
Of course, it would be beneficial to find as much overlap as possible
between aspects that both Android and iOS provide and just put them in
services. I'm sure iOS also provides a device identifier in one of its
apis, so that could be migrated into the services library already.

**Comment by abarth** [comment]
> find as much overlap as possible

Yeah, we should do that too.  We'll probably want to create a curated set of "flutter_services" that work across platforms and express our opinion about how to access these services.  Ideally, flutter_services would also be loosely coupled, etc.  The main issue is that building out that set of services requires forming opinions and making trade-offs between expressiveness, complexity, portability, etc.  Also, there's likely to be overlap with the service designs on FNL.  For example, we might just want to take the FNL protocols and make them work on other platforms if we think they're well-designed.

**Comment by jamesr** [comment]
Several of the current services are very Android-shaped and should likely be moved to android_services.  iOS specifically does *not* provide a device identifier.

**Comment by stevenroose** [comment]
How about VendorID?
https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIDevice_Class/#//apple_ref/occ/instp/UIDevice/identifierForVendor

**Comment by jason-simmons** [comment]
We recently added a feature that provides access to Android APIs and other Java libraries to Dart code in the Flutter environment.

You can use this to access the Android ID:

```
import 'dart:jni';

String getAndroidId() {
  JavaObject context = new JavaObject(JniApi.getApplicationContext());
  JavaClass secure = Java.getClass('android.provider.Settings\$Secure');
  return secure.getString(context.getContentResolver(), secure.ANDROID_ID);
}
```

**Comment by stevenroose** [comment]
I thought the one thing that makes Flutter great is the ability to avoid `if(platform=="android") doAndroid(); else doIOS();`...

**Comment by eseidelGoogle** [comment]
I hope platform abstractions are not the only things that make flutter great. :)

It certainly is possible to write a platform-abstracting service for a "device id" which returns secure.ANDROID_ID on Android and something else on iOS.  The flutter team hasn't written such a service, yet, but you or anyone else could (we don't yet make that as easy as we would like to, but are working on it). You see we've already written abstracting services like this for Keyboard, Network, Audio, etc.  However what @jason-simmons is demonstrating is that it's also in possible in flutter to punch below these services and access Java APIs (or soon Objective-C apis) directly.

We believe that different applications/authors will have different needs.  It's important not only that we provide good "first-party" services, but also that any "third party" can write their own, or as @jason-simmons showed above, be able to have quick-and-dirty access to the underlying platform sometimes w/o even writing a service.

**Comment by sethladd** [comment]
> Once all platforms Flutter supports support a feature, Flutter should implement it too and delegate to the platform's implementations.

For the cases where all platforms implement the same feature in the same way, this makes sense. But if two platforms implements the same feature with different capabilities, we don't want to provide a "lowest common denominator" API. If the platform can do it, we strive to make it possible for a developer to do it.

I can imagine an ecosystem of pub packages (which would come from the community) that provide unifying (and high-level) APIs across platforms. Flutter developers can opt-in to these higher-level APIs from the community, or choose to access lower-level (and potentially platform-specific) APIs. Or choose a combination! :)

**Comment by a14n** [comment]
@jason-simmons : I just tried to use your code snippet (with several versions of `sky_engine`). With `<0.0.88` my app displays only a black screen but I can see the deviceID in the output just before errors like `[ERROR:dart_error.cc(16)] '...xxx.dart': malformed type: line 5: type 'JavaObject' is not loaded`. With `0.0.88` I see this error message only once on the first start but my app seems to work.

1. Is it safe to start using it now or should I wait ?
1. How to configure the analyzer (ie. `_embedder.yaml`) to avoid warning about _Undefined JavaObject_ ?

**Comment by sethladd** [comment]
> How to configure the analyzer (ie. _embedder.yaml) to avoid warning about Undefined JavaObject ?

That sounds like a real warning we need to fix, and not ignore :) I guess that comes from dart:jni?

**Comment by a14n** [comment]
@sethladd yes it comes from dart:jni

**Comment by jason-simmons** [comment]
You will need sky_engine 0.0.88 or later

I just tried it with 0.0.88 and didn't see any errors.  The JavaObject class should be available if your code imports the 'dart:jni' library

**Comment by sethladd** [comment]
Thanks for confirming, @jason-simmons !

Flutter package (both master and alpha) right now is using sky_engine 0.0.87. Hang tight, we'll get everything upgraded :)

**Comment by abarth** [comment]
We've also got a plugin for this now.

**Comment by wilburt** [comment]
[This plugin](https://pub.dartlang.org/packages/device_info) does this now.

**Comment by matte5031** [comment]
@wilburt Hello! I don't get it, I installed the plugin device_info but can't see any way to get the android unique device id. Do you know how to get it using that plugin? any help or input appreciated, thanks!

**Comment by nvquangth** [comment]
@matte5031 I think you can use: 
`DeviceInfoPlugin deviceInfo = DeviceInfoPlugin();
    AndroidDeviceInfo androidInfo = await deviceInfo.androidInfo;
    print('uuid: ${androidInfo.androidId}');`

**Comment by Katekko** [comment]
Why the android ID will change if  the app is unistalled?? 
![image](https://user-images.githubusercontent.com/13632735/88398430-c495d480-cd9b-11ea-91f9-decb8d05ea9e.png)

Was a wrong typing?

**Comment by tapizquent** [comment]
> Why the android ID will change if the app is unistalled??
> ![image](https://user-images.githubusercontent.com/13632735/88398430-c495d480-cd9b-11ea-91f9-decb8d05ea9e.png)
> 
> Was a wrong typing?

Trying to understand this as well. Is there any way to get an ID that does not change when app is uninstalled and reinstalled?

**Comment by dhameliyamilan** [comment]
There is no any API to get deviceId but Here is Explained How to get it by step by step [How to Get DeviceId in Flutter with Example Source Code](https://fluttercorner.com/how-to-get-deviceid-in-flutter/)
